design/UsageValidationRules.md

```diff
@@ -0,0 +1,71 @@
+
+Resources in WebGPU can be used in different ways that are declared at resource creation time and interact with various validation rules.
+
+Buffers have the following usages (and commands inducing that usage):
+ - `VERTEX` for the `buffers` in calls to `WebGPUCommandEncoder.setVertexBuffers`
+ - `INDEX` for `buffer` in calls to `WebGPUCommandEncoder.setIndexBuffer`
+ - `INDIRECT` for `indirectBuffer` in calls to `WebGPUCommandEncode.{draw|drawIndexed|dispatch}Indirect`
+ - `UBO`, `RO_STORAGE` and `STORAGE` for buffers referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
+ - `TRANSFER_SRC` for buffers used as the copy source of various commands.
+ - `TRANSFER_DST` for buffers used as the copy destination of various commands.
+ - (Maybe [RO_]STORAGE_TEXEL and UNIFORM_TEXEL?)
+
+Textures have the following usages:
+ - `OUTPUT_ATTACHMENT` for the subresources referenced by `WebGPURenderPassDescriptor`
+ - `SAMPLED`, `RO_STORAGE` and `STORAGE` for subresources corresponding to the image views referenced by bindgroups passed to `setBindGroup`, with the usage corresponding to the binding's type.
+ - `TRANSFER_SRC` for textures used as the copy source of various commands.
+ - `TRANSFER_DST` for textures used as the copy destination of various commands.
+ 
+Read only usages are `VERTEX`, `INDEX`, `INDIRECT`, `UBO`, `RO_STORAGE`, `TRANSFER_SRC` and `SAMPLED`.
+
+Note that here we split `RO_STORAGE` and `STORAGE` compared to other graphics APIs and that creating a resource with `STORAGE` implies `RO_STORAGE`.
+
+## Render passes
+
+In render passes the only writable resources are textures used as `OUTPUT_ATTACHMENT` and resources used as `STORAGE`.
+
+To avoid data hazards the simplest validation rules would be to check for every subresource that it is used as either:
+ - `OUTPUT_ATTACHMENT`
+ - A combination of read-only usages
+ - A combination of `STORAGE` and `RO_STORAGE`
+
+If there is no usage of `STORAGE` then there are no data-races as everything is read-only, except `OUTPUT_ATTACHMENT` which is well-defined.
+`STORAGE` is an escape hatch and the order of reads and writes to the same `[RO_]STORAGE` resources is undefined.
+
+Implementations might need to track usage per mip-level per layer of a resource, if this is deemed too costly, we could only have two sub-resources tracked in textures: the part as `OUTPUT_ATTACHMENT` and the rest.
+This would mean for example that a texture couldn't be used as both  `[RO_]STORAGE` and a read-only usage inside a render pass.
+The state tracking required in implementation would become significantly simpler at the cost of flexibility for the applications.
+
+## Compute passes
+
+They are similar to render passes, every subresource (alternatively every resource) must be used as either:
+ - A combination of read-only usages
+ - A combination of `STORAGE` and `RO_STORAGE`
+
+By default implementation ensure there are no data-races between two dispatches.
```

Comment:
I'd propose to omit the synchronization section from here (as not directly relevant), since we are yet to submit "official" proposals on describing that